crates/lib/src/action/workingdata.rs

```diff
@@ -116,31 +120,98 @@ impl Default for WorkingData {
 ///
 /// The [`Action::outcome()`] method is the only way to set the outcome of the action, and it _must_
 /// be called before the handler returns.
-#[derive(Debug)]
 pub struct Action {
 	/// The collected events which triggered the action.
 	pub events: Arc<[Event]>,
-	pub(super) outcome: Arc<OnceCell<Outcome>>,
+	processes: Arc<[SupervisorId]>,
+	pub(crate) outcomes: Arc<Mutex<HashMap<SupervisorId, (Resolution, EventSet)>>>,
+}
+
+impl std::fmt::Debug for Action {
+	fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+		f.debug_struct("Action")
+			.field("events", &self.events)
+			.field("processes", &self.processes)
+			.finish()
+	}
 }
 
 impl Action {
-	pub(super) fn new(events: Arc<[Event]>) -> Self {
+	pub(crate) fn new(events: Arc<[Event]>, processes: Arc<[SupervisorId]>) -> Self {
 		Self {
 			events,
-			outcome: Default::default(),
+			processes,
+			outcomes: Default::default(),
 		}
 	}
 
-	/// Set the action's outcome.
+	/// Set the action's outcome for all [`Process`]es.
 	///
 	/// This takes `self` and `Action` is not `Clone`, so it's only possible to call it once.
 	/// Regardless, if you _do_ manage to call it twice, it will do nothing beyond the first call.
 	///
 	/// See the [`Action`] documentation about handlers to learn why it's a bad idea to clone or
 	/// send it elsewhere, and what kind of handlers you cannot use.
 	pub fn outcome(self, outcome: Outcome) {
-		self.outcome.set(outcome).ok();
+		let mut outcomes = tokio::task::block_in_place(|| self.outcomes.blocking_lock());
+		for process in self.processes.iter().copied().collect::<Vec<_>>() {
+			self.on_process_with_lock(&mut outcomes, process, outcome.clone(), EventSet::All);
+		}
 	}
+
+	/// Sets an [`Outcome`] for a single [`Process`].
+	pub async fn on_process(&self, process: SupervisorId, outcome: Outcome, set: EventSet) {
+		let mut guard = self.outcomes.lock().await;
+		self.on_process_with_lock(&mut guard, process, outcome, set);
+	}
+
+	// Used internally by on_process and outcome to insert into `outcomes`.
+	fn on_process_with_lock(
+		&self,
+		guard: &mut MutexGuard<'_, HashMap<SupervisorId, (Resolution, EventSet)>>,
+		process: SupervisorId,
+		outcome: Outcome,
+		set: EventSet,
+	) {
+		guard.insert(process, (Resolution::Apply(outcome), set));
+	}
+
+	/// Returns a snapshot of the `SupervisorId`s of the running `Process`es at creation of the
+	/// [`Action`].
+	pub fn current_processes(&self) -> &[SupervisorId] {
+		&self.processes
+	}
+
+	/// Starts a new [`Process`] with a provided [`Command`] and for a given [`EventSet`].
```

Comment:
Along with the rename of ProcessId, let's rename these methods and drop mention of "process", so it's not confusing that we're talking about a persistent process/command supervisor and not an OS-level process.

We can then rephrase as...

```rust
/// Starts a new supervised [`Command`].
///
/// This instantiates a new command supervisor, which manages the lifecycle
/// of a command within the watchexec instance, including stopping, starting
/// again, sending signals, and monitoring its aliveness.
///
/// You can control it by calling `command_outcome()` with the [`SupervisorId`]
/// this method returns. To destroy a supervisor, use `Outcome::End`.
///
/// For details on the `events` argument, see the documentation for
/// `command_outcome`.
///
/// Note that as this is async, your action handler must also be async. Calling
/// this method in a sync handler without `await`ing it will do nothing.
async fn start_command(&self, command: Command, events: EventSet) -> SupervisorId;
```

Reply:
yess, this is much nicer!!